### ISSUE TITLE
Every node.js at 100%CPU usage after v3.0.4 upgrade

### ISSUE THREAD

**Issue by magicbelette** [RocketChat/Rocket.Chat#17025]
### Description:

Migrating from 2.2.1 to 3.0.4, every node.js are stuck at 100%CPU usage and Rocket.Chat becomes unresponsive.

### Actual behavior:

The first peak at midnight is node.js instances restarting for 3.0.4 upgrade. Then in the morning it starts growing as users connect, Rocket.chat becomes (very) laggy and the load never decrease. Finally go back to Rocket.Chat 2.2.1, CPU load decreases and even if high load at the become, Rocket.Chat is still usable in 2.2.1.
![Capture d’écran_2020-03-26_12-28-32](https://user-images.githubusercontent.com/6803105/77642507-0df52500-6f5e-11ea-936a-ab81ff12b820.png)

### Server Setup Information:

- Version of Rocket.Chat Server: 3.0.4
- Operating System: Debian
- Deployment Method: tar
- Number of Running Instances: 99
- DB Replicaset Oplog: true
- NodeJS Version: v12.14.0
- MongoDB Version: 3.6

### Additional context

Node.js repartition :
host 1 with 8CPU : 7 node.js instances
host 1 with 64CPU : 62 node.js instances
host 1 with 32CPU : 30 node.js instances

That's maybe a bad idea to have a lot of node.js instances on the same server and keep small 4CPU servers with 3 node.js instances each ?

### Relevant logs:

A lot of differents errors but mainly, node.js instances seem to keep discovering each other :
```
Mar 25 18:33:41 m2-ariane-01 rocketchat3001[41025]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3001' to 'localhost:3002' with name notify-user not authorized
Mar 25 18:33:41 m2-ariane-01 rocketchat3001[41025]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3001' to 'localhost:3002' with name notify-user not authorized
Mar 25 18:33:41 m2-ariane-01 rocketchat3001[41025]: StreamBroadcast ➔ Auth.info Authorizing with localhost:3002
[...]
ar 25 18:33:51 m2-ariane-01 rocketchat3001[41025]: StreamBroadcast ➔ Auth.info broadcastAuth with localhost:3002 false
Mar 25 18:33:51 m2-ariane-01 rocketchat3001[41025]: StreamBroadcast ➔ Auth.info broadcastAuth with localhost:3002 false
[...]
Mar 25 19:07:53 m2-ariane-01 rocketchat3000[54349]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3000' to '10.167.124.87:3020' with name notify-logged not authorized
Mar 25 19:07:53 m2-ariane-01 rocketchat3000[54349]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3000' to '10.167.124.87:3016' with name notify-logged not authorized
Mar 25 19:07:53 m2-ariane-01 rocketchat3000[54349]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3000' to '10.167.124.87:3020' with name notify-user not authorized
Mar 25 19:07:53 m2-ariane-01 rocketchat3000[54349]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3000' to '10.167.124.87:3016' with name notify-user not authorized
[...]
Mar 25 18:45:27 m2-ariane-01 rocketchat3000[39494]: TIMEOUT QUERY OPERATION {
Mar 25 18:45:27 m2-ariane-01 rocketchat3000[39494]:   observeKey: '{"ordered":false,"collectionName":"users","selector":{"_id":"74Nqois8WyZ5CfBck"},"options":{"transform":null,"fields":{"profile":1,"username":1,"emails":1}}}',
Mar 25 18:45:27 m2-ariane-01 rocketchat3000[39494]:   writesToCommitWhenWeReachSteadyLength: 0,
Mar 25 18:45:27 m2-ariane-01 rocketchat3000[39494]:   cursorDescription: '{"collectionName":"users","selector":{"_id":"74Nqois8WyZ5CfBck"},"options":{"transform":null,"fields":{"profile":1,"username":1,"emails":1}}}'
Mar 25 18:45:27 m2-ariane-01 rocketchat3000[39494]: }
Mar 25 18:45:27 m2-ariane-01 rocketchat3000[39494]: Exception in setInterval callback: SwitchedToQuery {}
[...]
Mar 25 20:01:56 m2-ariane-01 rocketchat3000[61003]: (node:61003) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 exit listeners added to [process]. Use emitter.setMaxListeners() to increase limit
[...]
Mar 25 20:05:01 m2-ariane-01 rocketchat3045[66100]: StreamBroadcast ➔ Auth.info broadcastAuth with localhost:3002/ true
Mar 25 20:06:28 m2-ariane-01 rocketchat3002[61683]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3002' to 'localhost:3045' with name notify-user not authorized
```

**Comment by bbrauns** [comment]
@magicbelette would you share with us the number of online users you are serving with your setup? 

Currently we have one VM with 64 cores and around 20 RC instances for around 2000 online users. Yesterday we had an issue where all instances spiked to 100% CPU. We saw the same "Stream broadcast from xxx to xxx with name xxx not authorized" errors in our logs.

I hesitate to further increase the number or RC instances because I read that every instance has to contact every other instance which could potentially create bottlenecks. Version: 2.4.8

**Comment by magicbelette** [comment]
@bbrauns thanks for your answer.
We also serve 2000 online users with a peak to 7000. The problem appears with around 900 online users.
Yes you're right, every new instance annonces itself to mongodb in `instances` collection. Then every instances reads this collection to discover and broadcast to other instance.

**Comment by 564064202** [comment]
sir,I think,3.* is every unstable。Could you give a hand,but:
run docker-compose with multiple instances will meet a strange but ' MongoError: ns not found', 'Errors like this can cause oplog processing errors.'
https://github.com/RocketChat/Rocket.Chat/issues/17020

**Comment by magicbelette** [comment]
I noticed a difference between how instances are seen in admin page.
Rocket.Chat 2.2.1 : Auth and Connection status > Connected are unset
![rocketchat2 2 1](https://user-images.githubusercontent.com/6803105/77749622-9f2dcf80-7022-11ea-825f-9afa2702a969.png)

Rocket.Chat 3.0.4 : Auth and Connection status > Connected are set to true
![rocketchat3 0 4](https://user-images.githubusercontent.com/6803105/77749735-d00e0480-7022-11ea-91bd-5a7d6315f57e.png)

Maybe a new auth mecanism heavily load the streambroadcast process ?

**Comment by bbrauns** [comment]
2.4.8 reports auth and connection status:

![image](https://user-images.githubusercontent.com/6813310/77889564-5a967400-726e-11ea-8512-7bd2f54125d8.png)

**Comment by magicbelette** [comment]
OK, that's just a status report.

So.. Could it be due to this change @rodrigok ?
https://github.com/RocketChat/Rocket.Chat/commit/b95cb64742e67574ea4b5c419060a1d8fa5ee4b2#diff-067b50e42d48ca885e4b33cf54cf3124

**Comment by 564064202** [comment]
@magicbelette 
Could you give a hand,but:
run docker-compose with multiple instances will meet a strange but ' MongoError: ns not found', 'Errors like this can cause oplog processing errors.'
https://github.com/RocketChat/Rocket.Chat/issues/17020

**Comment by bbrauns** [comment]
We had another issue today. Starting a new instance leads to massive cpu usage of all the other rc instances. Starting multiple instances at once brings down the whole cluster.

**Comment by magicbelette** [comment]
> We had another issue today. Starting a new instance leads to massive cpu usage of all the other rc instances. Starting multiple instances at once brings down the whole cluster.

Got the same issue... Can you try to get back to this version ?
https://github.com/RocketChat/Rocket.Chat/commit/b95cb64742e67574ea4b5c419060a1d8fa5ee4b2

I'm setting up a new cluster to test with 64 instances.

**Comment by bbrauns** [comment]
Sorry I can't test in production. Customers are already mad. Why do you find https://github.com/RocketChat/Rocket.Chat/commit/b95cb64742e67574ea4b5c419060a1d8fa5ee4b2 suspicious?

**Comment by magicbelette** [comment]
> Sorry I can't test in production. Customers are already mad. Why do you find [b95cb64](https://github.com/RocketChat/Rocket.Chat/commit/b95cb64742e67574ea4b5c419060a1d8fa5ee4b2) suspicious?

Just because this was the only change somehow linked to streambroadcast... But it seems that the core team was quite busy to night:
https://github.com/RocketChat/Rocket.Chat/pull/17088
https://github.com/RocketChat/Rocket.Chat/pull/17089

As I saw a lot of 
`Mar 25 19:07:53 m2-ariane-01 rocketchat3000[54349]: server.js:204 StreamBroadcast ➔ Stream.error Stream broadcast from '10.167.124.86:3000' to '10.167.124.87:3016' with name notify-user not authorized
`
I'll try the last v3.0.9

**Comment by bbrauns** [comment]
Ok makes sense. But these changes aren't in v3.0.9 or am I wrong?
Please report your findings 😄

**Comment by magicbelette** [comment]
> We had another issue today. Starting a new instance leads to massive cpu usage of all the other rc instances. Starting multiple instances at once brings down the whole cluster.

Got the same issue... Can you try to get back to this version ?
https://github.com/RocketChat/Rocket.Chat/commit/b95cb64742e67574ea4b5c419060a1d8fa5ee4b2

I'm setting up a new cluster to test with 64 instances.

**Comment by magicbelette** [comment]
Those were included in 3.0.8
https://github.com/RocketChat/Rocket.Chat/releases/tag/3.0.8

**Comment by emikolajczak** [comment]
Hi @magicbelette
We have similar problems with high CPU load. Did you have a chance to test fixes in 3.0.8? Is there any 
improvement in performance?

**Comment by magicbelette** [comment]
Hi, the situation is far more comfortable in version 3.0.9. Plus, this one https://github.com/RocketChat/Rocket.Chat/pull/17115 should improve performance too
